data_structure/src/lesson7/GraphImpl.java

```diff
@@ -0,0 +1,186 @@
+        if (startIndex == -1) {
+            throw new IllegalArgumentException("Invalid start label: " + startLabel);
+        }
+
+        Stack<Vertex> stack = new Stack<>();
+        Vertex vertex = vertexList.get(startIndex);
+
+        visitVertex(stack, vertex);
+        while (!stack.isEmpty()) {
+            vertex = getNearUnvisitedVertex(stack.peek());
+            if (vertex != null) {
+                visitVertex(stack, vertex);
+            } else {
+                stack.pop();
+            }
+        }
+
+        resetVertexState();
+    }
+
+    private void resetVertexState() {
+        for (Vertex vertex : vertexList) {
+            vertex.setVisited(false);
+        }
+    }
+
+    private Vertex getNearUnvisitedVertex(Vertex current) {
+        int currentIndex = vertexList.indexOf(current);
+        for (int i = 0; i < getSize(); i++) {
+            if (adjMat[currentIndex][i] && !vertexList.get(i).isVisited()) {
+                return vertexList.get(i);
+            }
+        }
+        return null;
+    }
+
+    private void visitVertex(Stack<Vertex> stack, Vertex vertex) {
+        System.out.println(vertex.getLabel());
+        stack.push(vertex);
+        vertex.setVisited(true);
+    }
+
+    private void visitVertex(Queue<Vertex> queue, Vertex vertex) {
+        System.out.println(vertex.getLabel());
+        queue.add(vertex);
+        vertex.setVisited(true);
+    }
+
+    @Override
+    public void bfs(String startLabel) {
+        int startIndex = indexOf(startLabel);
+        if (startIndex == -1) {
+            throw new IllegalArgumentException("Invalid start label: " + startLabel);
+        }
+
+        Queue<Vertex> queue = new LinkedList<>();
+        Vertex vertex = vertexList.get(startIndex);
+
+        visitVertex(queue, vertex);
+        while (!queue.isEmpty()) {
+            vertex = getNearUnvisitedVertex(queue.peek());
+            if (vertex != null) {
+                visitVertex(queue, vertex);
+            } else {
+                queue.remove();
+            }
+        }
+
+        resetVertexState();
+    }
+
+    @Override
+    public Stack<String> searchPath(String firstTown, String lastTown) {
+        Stack<String> path = new Stack<>();
+        int firstIndex = indexOf(firstTown);
+        int lastIndex = indexOf(lastTown);
+        
+        if (firstIndex == -1 || lastIndex==-1){
+            throw new IllegalArgumentException();
+        }
+        
+        Queue<Vertex> queue = new ArrayDeque<>();
+        Vertex vertex = vertexList.get(firstIndex);
+        visitVertex(queue,vertex);
+        while (!queue.isEmpty()){
+            vertex = getNearUnvisitedVertex(queue.peek());
+            if(vertex == null){
+                queue.remove();
+            }
+            else {
+                visitVertex(queue,vertex);
+                vertex.setPreviousV(queue.peek());
+                if (vertex.getLabel().equals(lastTown)){
+                    path = setPath(vertex);
```

Comment:
```suggestion
                    path = setPath(vertex);
                    break;
```
поскольку нет смысла продолжать обход